Adding Data preparation + changing Evaluation model based on that

reyhanehjaafari opened the pull request:
_I have to change the evaluation model and also I have added the Preparation method to the project because I believe by backing the data and normalizing it. the result would be better. Let me explain the code in more detail;
**the full process for preparing the data is**:
Read text file and split into lines, split lines into pairs
Normalize text, filter by length and content
Make word lists from sentences in pairs_

def prepareData(lang1, lang2, reverse=False):
    input_lang, output_lang, pairs = readLangs(lang1, lang2, reverse)
    print("Read %s sentence pairs" % len(pairs))
    pairs = filterPairs(pairs)
    print("Trimmed to %s sentence pairs" % len(pairs))
    print("Counting words...")
    for pair in pairs:
        input_lang.addSentence(pair[0])
        output_lang.addSentence(pair[1])
    print("Counted words:")
    print(input_lang.name, input_lang.n_words)
    print(output_lang.name, output_lang.n_words)
    return input_lang, output_lang, pairs


input_lang, output_lang, pairs = prepareData('eng', 'fra', True)
print(random.choice(pairs))

**_Preparing Training Data_**
_To train, for each pair of words, we will need an input tensor (indexes of the words in the input sentence) and target tensor (indexes of the words in the target sentence). While creating these vectors we will append the EOS token to both sequences._
def indexesFromSentence(lang, sentence):
    return [lang.word2index[word] for word in sentence.split(' ')]


def tensorFromSentence(lang, sentence):
    indexes = indexesFromSentence(lang, sentence)
    indexes.append(EOS_token)
    return torch.tensor(indexes, dtype=torch.long, device=device).view(-1, 1)


def tensorsFromPair(pair):
    input_tensor = tensorFromSentence(input_lang, pair[0])
    target_tensor = tensorFromSentence(output_lang, pair[1])
    return (input_tensor, target_tensor)

**_Evaluation_**
_Every time it predicts a word we add it to the output string, and if it predicts the EOS token we stop there. We also store the decoder’s attention outputs for display later._
def evaluate(encoder, decoder, sentence, max_length=MAX_LENGTH):
    with torch.no_grad():
        input_tensor = tensorFromSentence(input_lang, sentence)
        input_length = input_tensor.size()[0]
        encoder_hidden = encoder.initHidden()

        encoder_outputs = torch.zeros(max_length, encoder.hidden_size, device=device)

        for ei in range(input_length):
            encoder_output, encoder_hidden = encoder(input_tensor[ei],
                                                     encoder_hidden)
            encoder_outputs[ei] += encoder_output[0, 0]

        decoder_input = torch.tensor([[SOS_token]], device=device)  # SOS

        decoder_hidden = encoder_hidden

        decoded_words = []
        decoder_attentions = torch.zeros(max_length, max_length)

        for di in range(max_length):
            decoder_output, decoder_hidden, decoder_attention = decoder(
                decoder_input, decoder_hidden, encoder_outputs)
            decoder_attentions[di] = decoder_attention.data
            topv, topi = decoder_output.data.topk(1)
            if topi.item() == EOS_token:
                decoded_words.append('<EOS>')
                break
            else:
                decoded_words.append(output_lang.index2word[topi.item()])

            decoder_input = topi.squeeze().detach()

        return decoded_words, decoder_attentions[:di + 1]

reyhanehjaafari committed:
'init'

reyhanehjaafari committed:
Merge branch 'main' of https://github.com/reyhanehjaafari/Prediction_Project

reyhanehjaafari committed:
Changing training Model

shirin1996 commented:
@reyhanehjaafari 
Thanks for your valuable contribution! It was insightful and I will take into consideration just one note that we will evaluate random sentences from the training set and print out the input, target, and output to make some subjective quality judgements! So I guess you should consider this in your new PR as well. 
Once yor code is updated, please submit your code in `.py` format to `Prediction_Project/src/ ` in the **new** branch.

reyhanehjaafari committed:
Add Random Evaluation

reyhanehjaafari commented:
@shirin1996 Thank you for your feedback. I have updated the pull request. I have considered a random evaluation process. I have added evaluate.py file in the src in the Add Random Evaluation commit

reyhanehjaafari committed:
Revert "Add Random Evaluation"

This reverts commit f4ff3d9707d646f51cc5206d2b13dad0161a71c0.

shirinyamani committed:
Delete Transformer.ipynb

shirinyamani reviewed (commented):
Thanks for your valuable contribution! It was insightful and I will take into consideration.
Just one note that we will evaluate random sentences from the training set and print out the input, target, and output to make some subjective quality judgements! So I guess you should consider this in your new PR as well.
Once yor code is updated, please submit your code and a new PR in .py format to Prediction_Project/src/ in the new branch. @reyhanehjaafari
